File: python-midonetclient/src/midonetclient/federation/midonet_binding.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4

# Copyright 2015 Midokura SARL
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging

from midonetclient import resource_base
from midonetclient.federation import vendor_media_type

LOG = logging.getLogger(__name__)


class MidonetBinding(resource_base.ResourceBase):
    media_type = vendor_media_type.MIDONET_BINDING_JSON

    def __init__(self, uri, dto, auth):
        super(MidonetBinding, self).__init__(uri, dto, auth)

    def get_id(self):
        return self.dto['id']

    def get_vtep(self):
        return self.dto['vtepId']

    def vtep(self, vtepId):
        self.dto['vtepId'] = vtepId
        return self

    def get_segment(self):
        return self.dto['segmentId']

    def segment(self, segmentId):
        self.dto['segmentId'] = segmentId
        return self

    def get_router_cidr(self):
        return self.dto['routerCidr']

    def router_cidr(self, routerCidr):
        self.dto['routerCidr'] = routerCidr
        return self

    def get_local_subnets(self):
        return self.dto['localSubnets']

    def local_subnets(self, localSubnets):
        self.dto['localSubnets'] = localSubnets
        return self

    def get_router_id(self):
        return self.dto['routerId']

    def router_id(self, routerId):
        self.dto['routerId'] = routerId
        return self

    # READ ONLY FIELDS
    def get_router_mac(self):
        return self.dto['routerMac']

    def get_router_port_id(self):
        return self.dto['routerPortId']

    def get_vtep_port_id(self):
        return self.dto['vtepPortId']

File: cluster/midonet-cluster/src/main/java/org/midonet/cluster/rest_api/models/federation/MidonetBinding.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.rest_api.models.federation;

import java.net.URI;
import java.util.List;
import java.util.UUID;

import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.xml.bind.annotation.XmlRootElement;

import com.fasterxml.jackson.annotation.JsonIgnore;

import org.midonet.cluster.data.ZoomClass;
import org.midonet.cluster.data.ZoomField;
import org.midonet.cluster.models.Federation;
import org.midonet.cluster.rest_api.models.UriResource;
import org.midonet.cluster.rest_api.validation.MessageProperty;
import org.midonet.cluster.util.IPSubnetUtil;
import org.midonet.cluster.util.UUIDUtil.Converter;
import org.midonet.packets.IPv4Subnet;
import org.midonet.packets.MAC;

@XmlRootElement
@ZoomClass(clazz = Federation.MidonetRouterBinding.class)
public class MidonetBinding extends UriResource {

    @ZoomField(name = "id", converter = Converter.class)
    public UUID id;

    @NotNull
    @ZoomField(name = "vtep_id", converter = Converter.class)
    public UUID vtepId;

    @NotNull
    @ZoomField(name = "segment_id", converter = Converter.class)
    public UUID segmentId;

    @NotNull
    @ZoomField(name = "router_id", converter = Converter.class)
    public UUID routerId;

    @ZoomField(name = "router_port_id", converter = Converter.class)
    public UUID routerPortId;

    @ZoomField(name = "vtep_port_id", converter = Converter.class)
    public UUID vtepPortId;

    @NotNull
    @Pattern(regexp = IPv4Subnet.IPV4_CIDR_PATTERN, message = MessageProperty.IP_ADDR_INVALID)
    @ZoomField(name = "router_cidr", converter = IPSubnetUtil.Converter.class)
    public String routerCidr;

    @NotNull
    //@Pattern(regexp = IPv4Subnet.IPV4_CIDR_PATTERN, message = MessageProperty.IP_ADDR_INVALID)
    @ZoomField(name = "local_subnets", converter = IPSubnetUtil.Converter.class)
    public List<String> localSubnets;

    @Pattern(regexp = MAC.regex, message = MessageProperty.IP_ADDR_INVALID)
    @ZoomField(name = "router_mac")
    public String routerMac;

    @Override
    public URI getUri() {
        return absoluteUri(Application.MIDONET_BINDINGS, id);
    }

    @JsonIgnore
    @Override
    public void create() {
        if (null == id) {
            id = UUID.randomUUID();
        }
        // TODO: generate a random MAC address
        // The router port and vtep port UUIDs will be selected by the
        // single-site API.

    }

    @JsonIgnore
    public void update(MidonetBinding from) {
        this.id = from.id;
    }
}
